File: hadoop-0.20.2/src/core/org/apache/hadoop/fs/shell/Command.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.hadoop.fs.shell;

import java.io.*;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.ipc.RemoteException;

/**
 * An abstract class for the execution of a file system command
 */
abstract public class Command extends Configured {
  protected String[] args;
  
  /** Constructor */
  protected Command(Configuration conf) {
    super(conf);
  }
  
  /** Return the command's name excluding the leading character - */
  abstract public String getCommandName();
  
  /** 
   * Execute the command on the input path
   * 
   * @param path the input path
   * @throws IOException if any error occurs
   */
  abstract protected void run(Path path) throws IOException;
  
  /** 
   * For each source path, execute the command
   * 
   * @return 0 if it runs successfully; -1 if it fails
   */
  public int runAll() {
    int exitCode = 0;
    for (String src : args) {
      try {
        Path srcPath = new Path(src);
        FileSystem fs = srcPath.getFileSystem(getConf());
        FileStatus[] statuses = fs.globStatus(srcPath);
        if (statuses == null) {
          System.err.println("Can not find listing for " + src);
          exitCode = -1;
        } else {
          for(FileStatus s : statuses) {
            run(s.getPath());
          }
        }
      } catch (RemoteException re) {
        exitCode = -1;
        String content = re.getLocalizedMessage();
        int eol = content.indexOf('\n');
        if (eol>=0) {
          content = content.substring(0, eol);
        }
        System.err.println(getCommandName() + ": " + content);
      } catch (IOException e) {
        exitCode = -1;
        System.err.println(getCommandName() + ": " + e.getLocalizedMessage());
      }
    }
    return exitCode;
  }
}

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Command.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.catalyst.plans.logical

import org.apache.spark.sql.catalyst.analysis.AnalysisContext
import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeSet}
import org.apache.spark.sql.catalyst.plans.QueryPlan
import org.apache.spark.sql.catalyst.trees.{BinaryLike, LeafLike, UnaryLike}
import org.apache.spark.sql.catalyst.trees.TreePattern.{COMMAND, TreePattern}

/**
 * A logical node that represents a non-query command to be executed by the system.  For example,
 * commands can be used by parsers to represent DDL operations.  Commands, unlike queries, are
 * eagerly executed.
 */
trait Command extends LogicalPlan {
  override def output: Seq[Attribute] = Seq.empty
  override def producedAttributes: AttributeSet = outputSet
  // Commands are eagerly executed. They will be converted to LocalRelation after the DataFrame
  // is created. That said, the statistics of a command is useless. Here we just return a dummy
  // statistics to avoid unnecessary statistics calculation of command's children.
  override def stats: Statistics = Statistics.DUMMY
  final override val nodePatterns: Seq[TreePattern] = Seq(COMMAND)
}

trait LeafCommand extends Command with LeafLike[LogicalPlan]
trait UnaryCommand extends Command with UnaryLike[LogicalPlan]
trait BinaryCommand extends Command with BinaryLike[LogicalPlan]

/**
 * A logical node that can be used for a command that requires its children to be only analyzed,
 * but not optimized. An example would be "create view": we don't need to optimize the view subtree
 * because we will just store the entire view text as is in the catalog.
 *
 * The way we do this is by setting the children to empty once the subtree is analyzed. This will
 * prevent the optimizer (or the analyzer from that point on) from traversing into the children.
 *
 * There's a corresponding rule
 * [[org.apache.spark.sql.catalyst.analysis.Analyzer.HandleSpecialCommand]] that marks these
 * commands analyzed.
 */
trait AnalysisOnlyCommand extends Command {
  val isAnalyzed: Boolean
  def childrenToAnalyze: Seq[LogicalPlan]
  override final def children: Seq[LogicalPlan] = if (isAnalyzed) Nil else childrenToAnalyze
  override def innerChildren: Seq[QueryPlan[_]] = if (isAnalyzed) childrenToAnalyze else Nil
  // After the analysis finished, we give the command a chance to update it's state based
  // on the `AnalysisContext`
  def markAsAnalyzed(analysisContext: AnalysisContext): LogicalPlan
}

/**
 * A logical node that does not expose its sub-nodes as children, but rather supervises them
 * in an implementation-defined manner.
 */
trait SupervisingCommand extends LeafCommand {
  /**
   * Transforms its supervised plan using `transformer` and returns a copy of `SupervisingCommand`
   */
  def withTransformedSupervisedPlan(transformer: LogicalPlan => LogicalPlan): LogicalPlan
}
